dotnet/src/InternalUtilities/src/System/TypeConverterFactory.cs

```diff
@@ -49,6 +51,41 @@ internal static class TypeConverterFactory
             return converter;
         }
 
-        return null;
+        // If a `TypeConverter` is not found for the custom type, then we will fallback to using 
+        // `JsonSerializer` to serialize/deserialize the value. This allows for custom types to 
+        // be supported by default.
+        return new JsonSerializationTypeConverter(type);
+    }
+
+    internal sealed class JsonSerializationTypeConverter : TypeConverter
+    {
+        private readonly Type _type;
+
+        public JsonSerializationTypeConverter(Type type)
+        {
+            this._type = type;
+        }
+
+        public override bool CanConvertFrom(ITypeDescriptorContext? context, Type sourceType) => sourceType == typeof(string);
+
+        public override object? ConvertFrom(ITypeDescriptorContext? context, CultureInfo? culture, object value)
+        {
+            if (value is not string)
+            {
+                throw new ArgumentOutOfRangeException(nameof(value), value, "Cannot convert from non-string value.");
+            }
+
+            return JsonSerializer.Deserialize(value.ToString()!, this._type);
```

Comment:
no need to call `ToString` method here. Just add variable name it the if block above - `if (value is not string json)`